tservice/src/seal.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Functions used for sealing and unsealing of secrets
+
+use alloc::{vec, vec::Vec};
+pub use core::ptr;
+pub use mc_sgx_core_types::{Error, Result};
+pub use mc_sgx_tservice_sys::sgx_calc_sealed_data_size;
+pub use mc_sgx_tservice_sys_types::sgx_sealed_data_t;
+pub use mc_sgx_tservice_types::{SealedData, UnsealedData};
+use mc_sgx_util::ResultInto;
+
+/// Handles the logic to seal data in an SGX enclave
+pub trait Seal<T> {
+    /// Returns the size needed to seal the `data`
+    ///
+    /// # Arguments
+    /// * `data` - The data to be sealed
+    fn sealed_data_size(&self) -> Result<usize>;
+
+    /// Seals the data
+    fn seal_data(&self) -> Result<SealedData<T>>;
+}
+
+impl<T: AsRef<[u8]>> Seal<Vec<u8>> for UnsealedData<T> {
+    /// Returns the size needed to seal the `data`
+    ///
+    /// # Arguments
+    /// * `data` - The data to be sealed
+    fn sealed_data_size(&self) -> Result<usize> {
+        let mac_length = match &self.mac {
+            Some(mac) => mac.as_ref().len() as u32,
+            None => 0,
+        };
+
+        let result = unsafe {
+            mc_sgx_tservice_sys::sgx_calc_sealed_data_size(
+                mac_length,
+                self.data.as_ref().len() as u32,
+            )
+        };
+
+        match result {
+            // Per the documentation, UINT32_MAX indicates an error
+            u32::MAX => Err(Error::Unexpected),
+            size => Ok(size as usize),
+        }
+    }
+
+    fn seal_data(&self) -> Result<SealedData<Vec<u8>>> {
+        let sealed_size = self.sealed_data_size()?;
+        let mut sealed_data = vec![0; sealed_size];
+
+        let (mac_pointer, mac_length) = match self.mac.as_ref() {
+            Some(text) => (text.as_ref().as_ptr(), text.as_ref().len() as u32),
+            None => (ptr::null(), 0),
+        };
+
+        unsafe {
+            mc_sgx_tservice_sys::sgx_seal_data(
+                mac_length,
+                mac_pointer,
+                self.data.as_ref().len() as u32,
+                self.data.as_ref().as_ptr(),
+                sealed_data.len() as u32,
+                sealed_data.as_mut_ptr() as *mut sgx_sealed_data_t,
+            )
+        }
+        .into_result()?;
+
+        SealedData::try_from(sealed_data).map_err(|_| Error::Unexpected)
```

Comment:
Not sure if we want a dedicated error type for Sealing/unsealing.
Or perhaps `FfiError` should be expanded to have an `mc_sgx_core_types::Error` variant

Reply:
no complaints so resolved